src/esp/gfx/GenericDrawable.cpp

```diff
@@ -38,31 +38,55 @@ void GenericDrawable::setLightSetup(const Mn::ResourceKey& resourceKey) {
   updateShader();
 }
 
-void GenericDrawable::draw(const Mn::Matrix4& transformationMatrix,
-                           Mn::SceneGraph::Camera3D& camera) {
-  updateShader();
-
+void GenericDrawable::updateShaderLightingParameters(
+    const Mn::Matrix4& transformationMatrix,
+    Mn::SceneGraph::Camera3D& camera) {
   const Mn::Matrix4 cameraMatrix = camera.cameraMatrix();
 
-  std::vector<Mn::Vector3> lightPositions;
+  // todo: move to MaterialHelper or elsewhere
+  std::vector<Mn::Vector4> lightPositions;
   lightPositions.reserve(lightSetup_->size());
-  std::vector<Mn::Color4> lightColors;
+  std::vector<Mn::Color3> lightColors;
   lightColors.reserve(lightSetup_->size());
+  std::vector<Mn::Color3> lightSpecularColors;
+  lightSpecularColors.reserve(lightSetup_->size());
+  constexpr float dummyRange = 10000;
+  std::vector<float> lightRanges(lightSetup_->size(), dummyRange);
+  const Mn::Color4 ambientLightColor = getAmbientLightColor(*lightSetup_);
 
   for (Mn::UnsignedInt i = 0; i < lightSetup_->size(); ++i) {
-    lightPositions.emplace_back(getLightPositionRelativeToCamera(
-        (*lightSetup_)[i], transformationMatrix, cameraMatrix));
-
-    lightColors.emplace_back((*lightSetup_)[i].color);
+    const auto& lightInfo = (*lightSetup_)[i];
+    lightPositions.emplace_back(Mn::Vector4(getLightPositionRelativeToCamera(
+        lightInfo, transformationMatrix, cameraMatrix)));
+
+    const auto& lightColor = (*lightSetup_)[i].color;
+    lightColors.emplace_back(lightColor);
+
+    // In general, a light's specular color should match its base color.
+    // However, negative lights have zero (black) specular.
+    constexpr Mn::Color3 blackColor(0.0, 0.0, 0.0);
+    bool isNegativeLight = lightColor.x() < 0;
```

Comment:
I suggest we treat negative lights as a kind of hidden feature. Don't advertise them in tutorials. It's likely they'll go away in a few months in favor of a proper ambient-occlusion system.